### PR TITLE
Added ability to disable Firebase setup via environment config

### PR DESCRIPTION
**Pull request by taras**
I want to disable Firebase in testing environment. I was able to replace the Firebase adapter but this initializer would cause problems because the Store and Model were still reopened. This code allows me to specify via the application's environment config to disable Firebase.

If you're willing to consider this change then I'll be happy to do what's necessary to make it mergable.

### CODE REVIEW

**Comment by tstirrat** [comment]
Hey! could you explain what happens when you don't disable the monkey patching? I wonder if we can solve this by being smarter in the other methods.

**Comment by taras** [comment]
What I'm doing is disabling Firebase in tests and use a custom JSONAPIAdapter that doesn't make any calls to backend API. For every test, I push fixed date using `store.push` and the adapter uses that data. 

When I don't disable monkey patching, I get errors when I try to push into the store.

**Comment by tstirrat** [comment]
Can you give me the text of the error? I'm trying to understand why and where it fails.

The patches to store check for the existence of the patched methods before calling them. This should prevent errors when the adapter has been swapped out at runtime.

**Comment by tstirrat** [comment]
Also, how are you disabling firebase, by switching in new adapters?

**Comment by taras** [comment]
Yes, I'm switching in a new adapter. 

I just realized that the problem could be fixed by checking that the adapter is an instance of Firebase adapter.

Like this
```
 // Monkeypatch the store until ED gives us a good way to listen to push events
    if (!DS.Store.prototype._emberfirePatched) {
      DS.Store.reopen({
        _emberfirePatched: true,
        push: function() {
          var record = this._super.apply(this, arguments);
          var modelName = record.constructor.modelName;
          var adapter = this.adapterFor(modelName);
          if (adapter.recordWasPushed) {
            adapter.recordWasPushed(this, modelName, record);
          }
          return record;
        },

        recordWillUnload: function(record) {
          var adapter = this.adapterFor(record.constructor.modelName);
          if (adapter instanceof FirebaseAdapter  && adapter.recordWillUnload) {
            adapter.recordWillUnload(this, record);
          }
        },

        recordWillDelete: function (record) {
          var adapter = this.adapterFor(record.constructor.modelName);
          if (adapter instanceof FirebaseAdapter && adapter.recordWillDelete) {
            adapter.recordWillDelete(this, record);
          }
        }
      });
    }

    if (!DS.Model.prototype._emberfirePatched) {
      DS.Model.reopen({
        _emberfirePatched: true,
        unloadRecord: function() {
          this.store.recordWillUnload(this);
          return this._super();
        },
        deleteRecord: function () {
          this.store.recordWillDelete(this);
          this._super();
        },

        ref: function () {
          if (this.__firebaseRef) {
            return this.__firebaseRef;
          }

          var adapter = this.store.adapterFor(this.constructor.modelName);
          if (adapter instanceof FirebaseAdapter && adapter._getRef) {
            return adapter._getRef(this.constructor, this.id);
          }
        }
      });
    }

    if (!DS.AdapterPopulatedRecordArray.prototype._emberfirePatched) {
      DS.AdapterPopulatedRecordArray.reopen({
        _emberfirePatched: true,
        willDestroy: function() {
          if (this.__firebaseCleanup) {
            this.__firebaseCleanup();
          }
          return this._super();
        }
      });
    }
```

**Comment by tstirrat** [comment]
The curious thing here is that only a **FirebaseAdapter** should have the property `adapter.recordWillUnload`, but from what you're saying, it seems like that conditional seems to resolve to `true` on your test adapter...

**Comment by tstirrat** [comment]
e.g. no base adapters have this method: https://github.com/emberjs/data/search?utf8=%E2%9C%93&q=recordWillUnload

or. are you extending your test adapter from **FirebaseAdapter**?

**Comment by taras** [comment]
I'm replacing FirebaseAdapter when running tests with a [TestAdapter](https://github.com/embersherpa/twitter/blob/master/tests/helpers/test-adapter.js) that's based on JSONAPIAdapter. 
Here is an [example of a test](https://github.com/embersherpa/twitter/blob/master/tests/acceptance/index-test.js#L14)

**Comment by tstirrat** [comment]
Weird. Very weird.

What is the value of `adapter.recordWillUnload` when you stop the debugger in `store. recordWillUnload`?

I tried to check out your repo to verify for myself but a few other tests are failing.

Can you let me know what I can do to reproduce this with the twitter repo?

**Comment by taras** [comment]
Ok, I created a branch that should make it easier to see the error https://github.com/embersherpa/twitter/tree/firebase-release

It's just using `ember-fire@1.4.7`.

_twitter_tests

It doesn't actually get to the adapter. It fails when calling `var adapter = this.adapterFor(record.constructor);` in the `push` method.

**Comment by tstirrat** [comment]
that last error is due to using 1.4.7 with ember-data 1.13.x (1.4.7 should be used with `<= beta.18`)

I'm working on the JSONAPI changes for 1.13.x compat (#267)

When I rolled forward emberfire to use `master` branch it fails on [this line](https://github.com/firebase/emberfire/blob/a9c75b4e289124598246fdffcf24745d6f4e2ddb/addon/initializers/emberfire.js#L30) because since the 1.13.x `store.push` can return an array of items and `record.constructor.modelName` is undefined.

I also tried rolling back ember data to `beta.18` but now your test adapter cant extend from `DS.JSONAPIAdapter` (doesnt exist)

Basically, emberfire still needs some fixes to be compatible with 1.13.x

Or, you should roll back to using ED `beta.18` (not 19)

I'll add this issue to #267 to make sure we capture the store.push changes
